pkg/networkservice/nsmonitor/client_test.go

```diff
@@ -0,0 +1,339 @@
+
+	client := chain.NewNetworkServiceClient(
+		metadata.NewClient(),
+		begin.NewClient(),
+		nsmonitor.NewClient(ctx, nsmonitor.WithSupplyMonitor(supplyMonitor)),
+	)
+
+	// no inodeURL in parameters
+	_, err := client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{
+				Parameters: map[string]string{},
+			},
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+
+	// no mechanism parameters
+	_, err = client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id:        uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{},
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+
+	// no mechanism
+	_, err = client.Request(ctx, &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+		},
+	})
+	require.NoError(t, err)
+	require.Empty(t, monitor.inodes)
+}
+
+func Test_Client_MonitorOnceSameConnectionId(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	monitor := testMonitor{}
+	supplyMonitor := func(c context.Context) nsmonitor.Monitor { return &monitor }
+
+	client := chain.NewNetworkServiceClient(
+		metadata.NewClient(),
+		begin.NewClient(),
+		nsmonitor.NewClient(ctx, nsmonitor.WithSupplyMonitor(supplyMonitor)),
+	)
+
+	request := &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{
+				Parameters: map[string]string{
+					common.InodeURL: testInode1,
+				},
+			},
+		},
+	}
+
+	_, err := client.Request(ctx, request)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(monitor.inodes))
+	require.Equal(t, testInode1, monitor.inodes[0])
+
+	_, err = client.Request(ctx, request)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(monitor.inodes))
+	require.Equal(t, testInode1, monitor.inodes[0])
+}
+
+func Test_Client_MonitorDifferentConnectionIds(t *testing.T) {
+	t.Cleanup(func() { goleak.VerifyNone(t) })
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+	defer cancel()
+
+	monitor := testMonitor{}
+	supplyMonitor := func(c context.Context) nsmonitor.Monitor { return &monitor }
+
+	client := chain.NewNetworkServiceClient(
+		metadata.NewClient(),
+		begin.NewClient(),
+		nsmonitor.NewClient(ctx, nsmonitor.WithSupplyMonitor(supplyMonitor)),
+	)
+
+	request1 := &networkservice.NetworkServiceRequest{
+		Connection: &networkservice.Connection{
+			Id: uuid.New().String(),
+			Mechanism: &networkservice.Mechanism{
+				Parameters: map[string]string{
+					common.InodeURL: testInode1,
+				},
+			},
+		},
+	}
+
+	_, err := client.Request(ctx, request1)
+	require.NoError(t, err)
+	require.Equal(t, monitor.inodes, []string{testInode1})
```

Comment:
The test can be unstable. Adding inodes for the test monitor is going in the goroutine. Please consider sync access. Moreover, add a check that monitor has handled watch operation to avoid race situations.